### ISSUE TITLE
Hardware keyboard issues(slow input and no Ctrl key)

### ISSUE THREAD

**Issue by iseikr** [pelya/xserver-xsdl#17]
Thank you developers for providing an excellent x server for android. My issue is that the X server handles input very slowly when entering different characters quickly in succession. If you for example spam 2 different keys repeatedly it will take seconds for the characters to appear on the screen(they appear after you stop typing). I was even able to reproduce this problem with an on-screen keyboard.
Another annoying problem is that the Ctrl is not registered by the x server.  xev  shows no events when pressing ctrl repeatedly or holding it.
I hope these issues can be fixed then it would feel like running a linux desktop natively on android.

**Comment by pelya** [comment]
Cannot reproduce - I've plugged USB keyboard into Nexus7, and XEV running on XSDL receives Ctrl/Alt/Shift keypresses just fine. Maybe your Bluetooth keyboard has some kind of mobile emulation mode, where it does not send Ctrl key immediately? Could you try to test it with PC or some other device or application?

XSDL redraws the screen as fast as it can, but it's limited to the network connection. Do you have slow WiFi? What device do you have?

**Comment by iseikr** [comment]
It is possible that Ctrl not grabbing problem is specific to my bluetooth keyboard, but other android applications such as JuiceSSH or ConnectBot grabs Ctrl fine. I will test with an usb keyboard as soon as  I can. In the meanwhile I was able to do work around the problem by binding my Ctrl key to a hardware key with "External keyboard helper" and then binding the hardware key to Ctrl in xserver-xsdl.

As for the slow input problem, I am testing this with a Nexus 10 and the Linux desktop is running on the same device in chroot so that should be extremely fast. Right now input is faster using vncserver on localhost but xserver-xsdl draws graphics much, much faster than vncserver. Input is fine in xserver-xsdl when holding a single key but when alternating between different keys it is slow. That indicates that the drawing speed is not the problem but it is rather a problem with how the application handles multiple keypresses or keys pressed quickly after each other(but it is just my guess).

**Comment by pelya** [comment]
Yes, it is slow, I don't really know why. VNC only redraws the parts of the
screen which were changed, but with XSDL, it's the X clients who control
what is redrawn, and standard edit fields seem to force redraw everything.
